src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Binders/EnumTypeModelBinder.cs

```diff
@@ -0,0 +1,121 @@
+
+using System;
+using System.ComponentModel;
+using System.Runtime.ExceptionServices;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.Mvc.ModelBinding.Binders
+{
+    /// <summary>
+    /// <see cref="IModelBinder"/> implementation to bind models for types deriving from <see cref="System.Enum"/>.
+    /// </summary>
+    public class EnumTypeModelBinder : IModelBinder
+    {
+        private readonly MvcOptions _mvcOptions;
+
+        public EnumTypeModelBinder(MvcOptions mvcOptions)
+        {
+            _mvcOptions = mvcOptions;
+        }
+
+        /// <inheritdoc />
+        public Task BindModelAsync(ModelBindingContext bindingContext)
+        {
+            if (bindingContext == null)
+            {
+                throw new ArgumentNullException(nameof(bindingContext));
+            }
+
+            var valueProviderResult = bindingContext.ValueProvider.GetValue(bindingContext.ModelName);
+            if (valueProviderResult == ValueProviderResult.None)
+            {
+                // no entry
+                return Task.CompletedTask;
+            }
+
+            bindingContext.ModelState.SetModelValue(bindingContext.ModelName, valueProviderResult);
+            var modelType = bindingContext.ModelMetadata.UnderlyingOrModelType;
+            var stringValue = valueProviderResult.FirstValue;
+
+            try
+            {
+                object model;
+                if (string.IsNullOrWhiteSpace(stringValue))
+                {
+                    model = null;
+                }
+                else
+                {
+                    var typeConverter = TypeDescriptor.GetConverter(modelType);
+                    model = typeConverter.ConvertFrom(
+                        context: null,
+                        culture: valueProviderResult.Culture,
+                        value: stringValue);
+                }
+
+                // Check if the converted value is indeed defined on the enum as EnumConverter converts 
+                // value to the backing type (ex: integer) and does not check if the value is defined on the enum.
+                if (!_mvcOptions.AllowBindingUndefinedValueToEnumType && model != null)
+                {
+                    var isFlagsEnum = modelType.IsDefined(typeof(FlagsAttribute), inherit: false);
+                    if (isFlagsEnum)
+                    {
+                        // From EnumDataTypeAttribute.cs in CoreFX
+                        var underlying = Convert.ChangeType(
+                            model,
+                            Enum.GetUnderlyingType(modelType),
+                            valueProviderResult.Culture).ToString();
+                        var converted = model.ToString();
+                        if (string.Equals(underlying, converted, StringComparison.OrdinalIgnoreCase))
+                        {
+                            model = null;
+                        }
+                    }
```

Comment:
Why do we need this? Won't the type converter have already done the right thing?

Reply:
Or is this just the check?